test/puppetlabs/puppetdb/command_test.clj

```diff
@@ -57,42 +57,38 @@
        (:name payload))
    payload])
 
-(defmacro test-msg-handler*
-  [command delay-var discard-var db & body]
-  `(tqueue/with-stockpile q#
-     (let [log-output#     (atom [])
-           delay# (call-counter)
-           dlo-dir# (fs/temp-dir "test-msg-handler-dlo")
-           dlo# (dlo/initialize (.toPath dlo-dir#)
-                                (:registry (new-metrics "puppetlabs.puppetdb.dlo"
-                                                        :jmx? false)))
-           handle-message# (mql/message-handler q# dlo#
-                                                delay#
-                                                #(process-command! % ~db))
-           cmd# ~command
-           cmdref# (-> (apply tqueue/store-command q# (unroll-old-command cmd#))
-                       (assoc :attempts (:attempts cmd#)))]
-       (try
-         (binding [*logger-factory* (atom-logger log-output#)]
-           (handle-message# cmdref#))
-         (let [~delay-var delay#
-               ~discard-var dlo-dir#]
-           ~@body
-           ;; Uncommenting this line can be very useful for debugging
-           ;; (println @log-output#)
-           )
-         (finally
-           (fs/delete-dir dlo-dir#))))))
-
-(defmacro test-msg-handler
-  "Converts `command` to JSON, runs it through a message-handler, and
-  then evaluates the `body` with `delay-var` bound to a call-counter,
-  and `discard-var` bound to the directory that contains failed
-  messages.  A history of failed attempts may be provided
-  via (:attempts `command`), which must be as a sequence of items
-  as created by (cons-attempt exception)."
-  [command delay-var discard-var & body]
-  `(test-msg-handler* ~command ~delay-var ~discard-var *db* ~@body))
+(defrecord CommandHandlerContext [message-handler command-chan dlo delay-pool response-chan q]
+  java.io.Closeable
+  (close [_]
+    (async/close! command-chan)
+    (async/close! response-chan)
+    (fs/delete-dir (:path dlo))
+    (#'overtone.at-at/shutdown-pool-now! @(:pool-atom delay-pool))))
+
+(defn call-with-message-handler [q]
+  (let [delay-pool (mk-pool)
+        command-chan (async/chan 10)
+        response-chan (async/chan 10)
```

Comment:
I imagine 10 here just because we know we never need more than that in the existing tests?